Cargo workspace2

xclaesse opened the pull request.

xclaesse commented:
@bonzini This is my very, very WIP for workspace support. The code is a total mess right now, but it finally passed configuration of the whole gst-plugins-rs tree! Over 400 subprojects generated.... https://paste.centos.org/view/6920dc55

bonzini commented:
Ok I will also do the same and apply some of your commits, so that the WIP parts are up to date and conflicts are not a problem. Hopefully we start getting some review soon. :)

bonzini commented:
Updated my branch(es). Among other things I renamed the conversion function to `_raw_to_dataclass`, switched the computed Meson version to lazy property.

I also added release notes for "cargo: Use -rs suffix only for rust ABI".

I would also like to reinforce that without unit tests for the dataclass conversion, this is not going anywhere. Some bugs that you have in your tree:

* changing the `default` feature to use default_factory is wrong, because you're dropping the feature if any other feature is defined

* as mentioned earlier, the version of the Cargo.lock file is an integer and not a string

Without tests it's also unclear what changes will need release notes. For example  I suspect that "cargo: When loading wraps multiple packages can have the same URL" does require them, but I am not sure.

I really would prefer it if you switched to https://github.com/bonzini/meson/commits/cargo-workspace2-for-xclaesse, since it also takes into account more of the review comments that you got for #12983. In any case I am very interested in moving this work forward, so I'll keep rebasing my branches and making them follow what you do as closely as possible.

xclaesse commented:
> I would also like to reinforce that without unit tests for the dataclass conversion, this is not going anywhere. 

We do have tests cases already, but I agree not enough.

> * changing the `default` feature to use default_factory is wrong, because you're dropping the feature if any other feature is defined

Thanks, reverted that change.

> * as mentioned earlier, the version of the Cargo.lock file is an integer and not a string

I missed that you mentioned that before, fixed.

> cargo: When loading wraps multiple packages can have the same URL

AFAIK, that's specific to workspace. Without workspace 1 subproject is 1 crate. For release notes, that's part of adding support for workspace.

> I really would prefer it if you switched to https://github.com/bonzini/meson/commits/cargo-workspace2-for-xclaesse

Could you rebase that branch on top of mine so I can more easily see what you're adding? I'm still very strongly against bringing back TypedDict.

bonzini commented:
> Could you rebase that branch on top of mine so I can more easily see what you're adding? I'm still very strongly against bringing back TypedDict.

I am not adding nothing just rebasing on top of #14672. I understand you don't want to bring back TypedDict but you already got pushback from 2.5 people (I am the 0.5).
